beacon/beacon.go

```diff
@@ -530,24 +527,29 @@ func (pd *ProtocolDriver) cleanupEpoch(epoch types.EpochID) {
 }
 
 // listens to new layers.
-func (pd *ProtocolDriver) listenLayers(ctx context.Context) {
+func (pd *ProtocolDriver) listenEpochs(ctx context.Context) {
 	pd.logger.With().Info("starting listening layers")
 
+	currentEpoch := pd.clock.GetCurrentLayer().GetEpoch()
+	layer := currentEpoch.Add(1).FirstLayer()
 	for {
 		select {
 		case <-pd.ctx.Done():
 			return
-		case layer := <-pd.layerTicker:
-			pd.logger.With().Debug("received tick", layer)
-			if layer.FirstInEpoch() {
-				epoch := layer.GetEpoch()
-				pd.setProposalTimeForNextEpoch()
-				pd.logger.With().Info("first layer in epoch", layer, epoch)
-				pd.eg.Go(func() error {
-					_ = pd.onNewEpoch(ctx, epoch)
-					return nil
-				})
+		case <-pd.clock.AwaitLayer(layer):
+			current := pd.clock.GetCurrentLayer()
+			if current.Before(layer) {
+				pd.logger.With().Info("time sync detected, realigning Beacon")
+				continue
 			}
+			layer = current.GetEpoch().Add(1).FirstLayer()
+			epoch := current.GetEpoch()
```

Comment:
current may not be the first layer of this epoch in the case of delayed tick. i don't think the node should participate in the beacon protocol when it's late.

Reply:
I updated the code to skip the epoch if the tick is late.